[JIT] Document torch.jit.interface

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109356

Good option for replacing "Callable" types; we should document it so
it's searchable.

davidberard98 committed:
[JIT] Document torch.jit.interface

Good option for replacing "Callable" types; we should document it so
it's searchable.

[ghstack-poisoned]

davidberard98 commented:
reminder to myself to check https://docs-preview.pytorch.org/pytorch/pytorch/109356/generated/torch.jit.interface.html once the build finishes

davidberard98 committed:
add to docs/source/jit.rst? on "[JIT] Document torch.jit.interface"


Good option for replacing "Callable" types; we should document it so
it's searchable.

[ghstack-poisoned]

eellison reviewed (approved):
nice!

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -1468,6 +1468,52 @@ def _check_directly_compile_overloaded(obj):
+    r"""
+    This decorator can be used to define an interface that can be used to annotate
+    classes or modules of different types. This can be used for to annotate a submodule
+    or attribute class that could have different values, or which could be swapped
```
```suggestion
    or attribute class that could have different types that implement the same interface, or which could be swapped
```

davidberard98 committed:
Update on "[JIT] Document torch.jit.interface"


Good option for replacing "Callable" types; we should document it so
it's searchable.

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge

davidberard98 commented:
@pytorchbot merge -i
